src/plan/generational/barrier.rs

```diff
@@ -0,0 +1,158 @@
+use crate::util::*;
+use crate::vm::{ObjectModel, VMBinding};
+use crate::MMTK;
+
+use super::gc_work::GenNurseryProcessEdges;
+use super::gc_work::ProcessArrayCopyModBuf;
+use super::gc_work::ProcessModBuf;
+use super::global::Gen;
+
+/// Object barrier for generational collection
+pub struct GenObjectBarrier<VM: VMBinding> {
+    /// MMTk instance
+    mmtk: &'static MMTK<VM>,
+    /// Generational plan
+    gen: &'static Gen<VM>,
+    /// Object modbuf. Contains a list of objects that may contain pointers to the nursery space.
+    modbuf: Vec<ObjectReference>,
+    /// Array-copy modbuf. Contains a list of sub-arrays or array slices that may contain pointers to the nursery space.
+    arraycopy_modbuf: Vec<(Address, usize)>,
+    /// Max size of the modbuf(s).
+    capacity: usize,
+}
+
+impl<VM: VMBinding> GenObjectBarrier<VM> {
+    #[allow(unused)]
+    pub fn new(mmtk: &'static MMTK<VM>, gen: &'static Gen<VM>, capacity: usize) -> Self {
+        Self {
+            mmtk,
+            gen,
+            modbuf: vec![],
+            arraycopy_modbuf: vec![],
+            capacity,
+        }
+    }
+
+    /// Attepmt to atomically log an object.
+    /// Returns true if the object is not logged previously.
+    #[inline(always)]
+    fn object_is_unlogged(&self, object: ObjectReference) -> bool {
+        load_metadata::<VM>(&VM::VMObjectModel::GLOBAL_LOG_BIT_SPEC, object, None, None) != 0
+    }
+
+    /// Attepmt to atomically log an object.
+    /// Returns true if the object is not logged previously.
+    #[inline(always)]
+    fn log_object(&self, object: ObjectReference) -> bool {
+        loop {
+            let old_value = load_metadata::<VM>(
+                &VM::VMObjectModel::GLOBAL_LOG_BIT_SPEC,
+                object,
+                None,
+                Some(Ordering::SeqCst),
+            );
+            if old_value == 0 {
+                return false;
+            }
+            if compare_exchange_metadata::<VM>(
+                &VM::VMObjectModel::GLOBAL_LOG_BIT_SPEC,
+                object,
+                1,
+                0,
+                None,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                return true;
+            }
+        }
+    }
+
+    /// object barrier slow-path call
+    #[inline(never)]
+    fn gen_object_reference_write_slow_no_inline(&mut self, src: ObjectReference) {
+        self.gen_object_reference_write_slow(src)
+    }
+
+    /// object barrier slow-path call.
+    /// Can be called directly by bindings to for a specialized slow-path call
+    #[inline(always)]
+    pub fn gen_object_reference_write_slow(&mut self, src: ObjectReference) {
+        // Log and enqueue the object if it is unlogged
+        if self.log_object(src) {
+            // enqueue the object
+            self.modbuf.push(src);
+            // the buffer is full?
+            if self.modbuf.len() >= self.capacity {
+                self.flush();
```

Comment:
Also you could have a type of `ModBuffer`. The pattern of `push/check-length/flush` currently appears many times in the code.

Reply:
I switched to using `VectorObjectQueue` for the mod buffers. I also had to rename `VectorObjectQueue` to `VectorQueue<T>`, so it can be used as an array-slice queue as well.

The `push/check-length/flush` code is simplified, and most of them are delegated to  `VectorObjectQueue`.